packages/application-shell/src/components/navbar/navbar.js

```diff
@@ -247,39 +248,55 @@ export const MenuItemDivider = () => (
 
 MenuItemDivider.displayName = 'MenuItemDivider';
 
-// This component basically just wraps the `<RestrictedByPermissions>`
+// This component wraps the `<ApplicationContext>`, `<RestrictedByPermissions>`
 // and the `<ToggleFeature>` components. However, it's necessary to have it as
 // the `<ToggleFeature>` wrapper should be rendered only if the `featureToggle`
 // prop is defined. This is because `<ToggleFeature>` will not render any
 // children if the flag is missing/not found.
-export const ToggledWithPermissions = props => {
-  const permissionsWrapper =
-    props.permissions.length > 0 ? (
-      <RestrictedByPermissions
-        permissions={props.permissions}
-        // Always check that some of the given permissions match.
-        shouldMatchSomePermissions={true}
-      >
-        {props.children}
-      </RestrictedByPermissions>
-    ) : (
-      props.children
-    );
-  return props.featureToggle ? (
-    <ToggleFeature flag={props.featureToggle}>
-      {permissionsWrapper}
-    </ToggleFeature>
-  ) : (
-    permissionsWrapper
-  );
-};
-ToggledWithPermissions.displayName = 'ToggledWithPermissions';
-ToggledWithPermissions.propTypes = {
+export const RestrictedMenuItem = props => (
```

Comment:
I renamed this but this whole `<NavBar />` starts getting a bit worried. I didn't dare to touch too much as it looks like it would fall apart. I think it's a prime candidate to refactoring from code and to use rtl but I am not sure how much design would like to change things too. Hence I just keep things as is.

Reply:
That's a good point. If you take a look at the whole file (at least for me) is pretty hard to follow because there is something that points to the top of the file another thing that uses stuff from the bottom so ...

I become a super fan of declarative stuff and here gives the feeling of us trying to automate the rendering somehow. This component is "finite" at some point. I got for example that the custom application is more dynamic but the rest I don't think so.

```
Menu
 MenuEntry (products)
   MenuItem (add product)
   MenuItem (list)
 MenuEntry (customers)
   MenuItem (add customer)
````

Surely worths a refactor 😋

Reply:
A rewrite of the navbar is already planned, I just need to finalise some things with Sven first before we start the refactoring.